### PR TITLE
feat: achievements repo

### PR DESCRIPTION
**Pull request by yo-ru**

### CODE REVIEW

**Comment by yo-ru** [comment]
was a little skeptical about this pr, but it may prove useful in the future.

ie.
- custom achievement creation
- possible updating of existing achievement conditions
- etc.

LGTM, will test rq.

File: app/repositories/achievements.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    file: str,
+    name: str,
+    desc: str,
+    cond: str,
+) -> dict[str, Any]:
+    """Create a new achievement."""
+    query = """\
+        INSERT INTO achievements (file, name, desc, cond)
+             VALUES (:file, :name, :desc, :cond)
+
```
Comment (cmyui) [inline app/repositories/achievements.py:36]:
```suggestion
```

File: app/repositories/achievements.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    rec = await app.state.services.database.fetch_one(query, params)
+    return dict(rec) if rec is not None else None
+
+
+async def fetch_count() -> int:
+    """Fetch the number of achievements."""
+    query = """\
+        SELECT COUNT(*) as count
```
Comment (cmyui) [inline app/repositories/achievements.py:85]:
```suggestion
        SELECT COUNT(*) AS count
```

File: app/repositories/achievements.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+) -> list[dict[str, Any]]:
+    """Fetch a list of achievements."""
+    query = f"""\
+        SELECT {READ_PARAMS}
+          FROM achievements
+    """
+    params = {}
+
+    if page is not None and page_size is not None:
+        query += " LIMIT :page_size OFFSET :offset"
```
Comment (cmyui) [inline app/repositories/achievements.py:107]: format
```suggestion
        query += " LIMIT :page_size OFFSET :offset"
```

File: app/repositories/achievements.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+
+async def delete(
+    id: int,
+) -> Optional[dict[str, Any]]:
+    """Delete an existing achievement."""
+    query = f"""\
+        SELECT {READ_PARAMS}
+            FROM achievements
+              WHERE id = :id
```
Comment (cmyui) [inline app/repositories/achievements.py:160]:
```suggestion
        SELECT {READ_PARAMS}
          FROM achievements
         WHERE id = :id
```

File: app/repositories/achievements.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        "id": id,
+    }
+    rec = await app.state.services.database.fetch_one(query, params)
+    if rec is None:
+        return None
+
+    query = """\
+        DELETE FROM achievements
+         WHERE id = :id
+
```
Comment (cmyui) [inline app/repositories/achievements.py:172]:
```suggestion
```

File: app/repositories/achievements.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+) -> Optional[dict[str, Any]]:
+    """Update an existing achievement."""
+    query = """\
+        UPDATE achievements
+           SET file = COALESCE(:file, file),
+               name = COALESCE(:name, name),
+               desc = COALESCE(:desc, desc),
+               cond = COALESCE(:cond, cond)
+         WHERE id = :id
+
```
Comment (cmyui) [inline app/repositories/achievements.py:132]:
```suggestion
```

**Review by cmyui — APPROVED** [review]

**Pull request merged**